nav2-usability/scenario-code/skipping-stacks/router.dart

```diff
@@ -0,0 +1,503 @@
+
+  BookRouterDelegate() : navigatorKey = GlobalKey<NavigatorState>() {
+    _appState.addListener(() => notifyListeners());
+  }
+
+  @override
+  AppRoutePath get currentConfiguration {
+    final selectedBookId = _appState.selectedBookId;
+    final selectedAuthorId = _appState.selectedAuthorId;
+    final showAuthorsPage = _appState.showAuthorsPage;
+
+    if (showAuthorsPage) {
+      return AuthorsRoutePath();
+    }
+
+    if (selectedAuthorId != null) {
+      return AuthorRoutePath(selectedAuthorId);
+    }
+
+    if (selectedBookId != null) {
+      return BookRoutePath(selectedBookId);
+    }
+
+    return AppRoutePath();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final selectedBook = _appState.selectedBook;
+    final selectedAuthor = _appState.selectedAuthor;
+    return Navigator(
+      key: navigatorKey,
+      pages: [
+        if (selectedAuthor != null) ...[
+          MaterialPage(
+            key: ValueKey('AuthorsListPage'),
+            child: AuthorsListScreen(
+              authors: authors,
+              onTapped: _handleAuthorTapped,
+              onGoToBooksTapped: _handleGoToBooksTapped,
+            ),
+          ),
+          AuthorDetailPage(
+            author: selectedAuthor,
+          ),
+        ] else if (selectedBook != null) ...[
+          MaterialPage(
+            key: ValueKey('BooksListPage'),
+            child: BooksListScreen(
+              books: books,
+              onTapped: _handleBookTapped,
+            ),
+          ),
+          BookDetailsPage(
+            book: selectedBook,
+            onAuthorTapped: _handleAuthorTapped,
+          )
+        ] else if (_appState.showAuthorsPage)
+          MaterialPage(
+            key: ValueKey('AuthorsListPage'),
+            child: AuthorsListScreen(
+              authors: authors,
+              onTapped: _handleAuthorTapped,
+              onGoToBooksTapped: _handleGoToBooksTapped,
+            ),
+          )
+        else
+          MaterialPage(
+            key: ValueKey('BooksListPage'),
+            child: BooksListScreen(
+              books: books,
+              onTapped: _handleBookTapped,
+            ),
+          ),
+      ],
+      onPopPage: (route, result) {
+        if (!route.didPop(result)) {
+          return false;
+        }
+
+        if (_appState.selectedAuthor != null) {
+          _appState.showAuthorsPage = true;
+        }
+
+        _appState.clearSelectedBook();
+        _appState.clearSelectedAuthor();
+
+        return true;
+      },
+    );
+  }
+
+  @override
+  Future<void> setNewRoutePath(AppRoutePath path) async {
+    if (path is AuthorRoutePath) {
+      _appState.selectedAuthorId = path.id;
+    } else if (path is AuthorsRoutePath) {
+      _appState.showAuthorsPage = true;
+    } else if (path is BookRoutePath) {
+      _appState.selectedBookId = path.id;
+    } else {
+      _appState.clearSelectedBook();
+      _appState.clearSelectedAuthor();
```

Comment:
This may be a problem if we currently on AuthorsRoutePath and receives AppRoutePath

Reply:
Added `_appState.isViewingAuthorsPage = false` to make sure it takes them to the home page.